File: 0121-best-time-to-buy-and-sell-stock/solution.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        const int n = prices.size();
        int res = 0 ; 
        
        int check = 0;
        for(int i = n - 2 ; i >= 0 ; i-- ){
            check = max(check , prices[i+1]);
            res = max(res , check - prices[i]);
        }
        
        return res ; 
    }
};

File: 0121-best-time-to-buy-and-sell-stock/solution.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length ;
        int mx = prices[n-1];

        int res = 0 ;
        for(int i = n - 2 ; i >= 0 ; i--){
            res = Math.max(res, mx - prices[i]);
            mx = Math.max(mx,prices[i]);
        }

        return res ;
    }
}
